roxie/ccd/ccdserver.cpp

```diff
@@ -784,65 +786,60 @@ class CRoxieServerActivityFactory : public CRoxieServerActivityFactoryBase
-        : CRoxieServerActivityFactory(_id, _subgraphId, _queryFactory, _helperFactory, _kind, _graphNode)
-    {
-    }
+    mutable std::vector<CRuntimeStatisticCollection> edgestatsArray;
 
-    ~CRoxieServerMultiOutputFactory()
+    CRoxieServerMultiOutputFactory(unsigned _id, unsigned _subgraphId, IQueryFactory &_queryFactory, HelperFactory *_helperFactory, ThorActivityKind _kind, IPropertyTree &_graphNode, const StatisticsMapping &_factoryStats)
+        : CRoxieServerActivityFactory(_id, _subgraphId, _queryFactory, _helperFactory, _kind, _graphNode, _factoryStats)
     {
-        delete [] processedArray;
-        delete [] startedArray;
     }
 
     void setNumOutputs(unsigned num)
     {
         numOutputs = num;
         if (!num)
             num = 1; // Even sink activities like to track how many records they process
-        processedArray = new RelaxedAtomic<__int64>[num];
-        startedArray = new RelaxedAtomic<__int64>[num];
+        edgestatsArray.reserve(num);
         for (unsigned i = 0; i < num; i++)
         {
-            processedArray[i] = 0;
-            startedArray[i] = 0;
+            edgestatsArray.push_back(CRuntimeStatisticCollection(edgeStatisticsX));
         }
     }
 
-    virtual void getEdgeProgressInfo(unsigned idx, IPropertyTree &edge) const
-    {
-        assertex(numOutputs ? idx < numOutputs : idx==0);
-        putStatsValue(&edge, "NumRowsProcessed", "sum", processedArray[idx]);
-        putStatsValue(&edge, "NumStarts", "sum", startedArray[idx]);
-    }
-
     virtual void resetNodeProgressInfo()
     {
         CRoxieServerActivityFactory::resetNodeProgressInfo();
         for (unsigned i = 0; i < numOutputs; i++)
         {
-            processedArray[i] = 0;
-            startedArray[i] = 0;
+            edgestatsArray[i].reset();
         }
     }
 
+    virtual void gatherEdgeStats(IStatisticGatherer &builder, int channel, bool reset) const override
+    {
+        // Edges here (other than for splitter) are a bit odd - there may not be a corresponding edge in the graph to
+        // record the stats against, and if there is the scope string for it is probably not what we expect, as these
+        // edges are artificially added in response to tying spills to reads, etc.
+        // So ignore them for now
+    }
+
     virtual void noteProcessed(unsigned idx, unsigned _processed) const
      {
          assertex(numOutputs ? idx < numOutputs : idx==0);
-         processedArray[idx] += _processed;
+         edgestatsArray[idx].addStatistic(StNumRowsProcessed, _processed);
      }
 
     virtual void noteStarted(unsigned idx) const
     {
         assertex(numOutputs ? idx < numOutputs : idx==0);
-        startedArray[idx]++;
+        edgestatsArray[idx].addStatistic(StNumStarts, 1);
+        edgestatsArray[idx].addStatistic(StNumStops, 1);
```

Comment:
not sure about incrementing numstops here - it should at least have a comment to indicate why it is ok

Reply:
We don't really track stops at factory level. Added a comment.